blog/2020-07-13-alpine-python.md

```diff
@@ -0,0 +1,156 @@
+
+I read often: Python is slower if they use musl as default C library. This fact is not 100% correct. If both Python runtime was compiled with the same GCC and with `-O3`, the glibc variant is a bit faster in the lab benchmark, but in the real world, it makes not a difference. Alpine compile it for `-Os` and most distro with `-O2`, which gives the often written difference between the Python runtime interpreters, but with the same compiler optimizing, musl based Python runtimes have no negative side effect.
+
+But there is a game-changer which makes the musl one more useful as the glibc based runtime. It is the memory allocator [jemalloc](http://jemalloc.net/), a general-purpose malloc implementation that emphasizes fragmentation avoidance and scalable concurrency support. There is an interesting effect which I found on some blogpost about Rust. There were some developers who saw that musl is much faster as glibc with jemalloc, but glibc is slower with jemalloc as without. For sure, the benefit with glibc and jemalloc is not the speed as they optimize memory management, but musl get both benefits. While the difference between pure musl and glibc can be ignored, the difference between musl + jemalloc and glibc are substantial (with disabled GCC memory allocator built-in optimization). Yes, jemalloc is compatible today with musl, and there was a time which it was not.
+
+## Compiler
+
+How do you compile Python is also essential. There were at least statements from Fedora or Redhat about disable `semantic-interposition` to become a high-performance boost. I was not able to reproduce this on GCC 9.3.0, but I also saw no adverse side effects. I can recommend disabling the semantics like the built-in allocator optimization and link jemalloc at build time. I will also recommend using the `-O3` optimization. We never saw an issue with this aggressive optimization on our target platforms. I need to say, unlike the distro Python runtime interpreters, we don't need to run everywhere. So we can use the `--enable-optimizations` without any overwrite and add more flags. I can say today, PGO/LTO/O3 make Python faster and at least work on our target CPUs.
+
+## Python packages
+
+Yes, Alpine indeed has no manylinux compatibility with musl. If you don't cache your builds, it needs every time compiling the C extensions on that packages. This process would take time. More time if you also crossbuild with Qemu for different CPU architectures. You can not get precompiled binaries from PyPi. I don't see a loose there because the packages are mostly also not optimized for the target system. We created our wheel [index](https://wheels.home-assistant.io/) and [backend]( https://github.com/home-assistant/wheels) to compile and hold it up to date over CI agents. For sure, not every project can take this effort, but also not all repository haver over 1k packages for each CPU architecture, and the build time of the docker file is not so important at all.
```

Comment:
```suggestion
Alpine Linux has no [manylinux](https://github.com/pypa/manylinux) compatibility with musl. If you don't cache your builds then it's required to compiling the C extensions for that packages every time. This process takes time. More time if you also crossbuild with QEMU for different CPU architectures. You can not get precompiled binaries from PyPi. I don't see a downside here because the packages are mostly not optimized for the target systems. We created our wheel [index](https://wheels.home-assistant.io/) and [backend]( https://github.com/home-assistant/wheels) to compile and keep it up to date with CI agents. For sure, not every project can take this effort, but also not all repository have over 1000 packages for each CPU architecture. The build time of the Docker file is not the limiting factor at all.
```